In [5]:
import requests
import pandas as pd
from bs4 import BeautifulSoup
import numpy as np
import time
import json
import re

In [4]:
from selenium.webdriver.edge import service
from selenium import webdriver
from selenium.webdriver.common.by import By
from collections import Counter
from nltk.corpus import stopwords
from selenium.webdriver.common.action_chains import ActionChains

ModuleNotFoundError: No module named 'selenium'

## Scraping the list of XDA Community apps

Scraping first page:

In [3]:
link_start = "https://forum.xda-developers.com/f/android-apps-and-games.530/"

In [4]:
r = requests.get(link_start)
soup = BeautifulSoup(r.text)
soup.find(class_="structItem-title")
links=[]
for link in soup.find_all(class_="structItem-title"):
    url = link.find("a").get("href")
    title = link.find("a").get_text()
    urls_dict = {"url": url,"title": title}
    links.append(urls_dict)

Drop sticky threads:

In [5]:
app_links = links[4:].copy()

In [6]:
app_links

[{'url': '/t/app-6-0-compass-of-chance.4440211/',
  'title': '[APP][6.0+]Compass of Chance'},
 {'url': '/t/app-5-0-fairemail-fully-featured-open-source-privacy-oriented-email-app.3824168/',
  'title': '[APP][5.0+] FairEmail - Fully featured, open source, privacy oriented email app'},
 {'url': '/t/app-all-root-solutions-6-0-viper4android-fx-2-7.3774651/',
  'title': '[APP][ALL ROOT SOLUTIONS][6.0+]ViPER4Android FX 2.7'},
 {'url': '/t/xfinity-tv-mod-android-box-hdmi-output.3535881/',
  'title': 'Xfinity TV Mod Android Box (HDMI output)'},
 {'url': '/t/app-6-0-netguard-no-root-firewall.3233012/',
  'title': '[APP][6.0+] NetGuard - No-root firewall'},
 {'url': '/t/app-4-1-kiwi-browser-chromium-adblock-caf-night-mode-2021-07-29-active.3797252/',
  'title': '[APP][4.1+] Kiwi Browser? Chromium + Adblock + CAF + Night Mode 2021-07-29|Active'},
 {'url': '/t/guide-multilingo-keyboard.4064431/',
  'title': '[GUIDE] MultiLingO Keyboard'},
 {'url': '/t/app-2-2-root-wifi-reaver-gui-for-android.24568

Scraping after the first page:

In [7]:
for page in range(2, 2135):
    page = link_start + 'page-' + str(page)
    r = requests.get(page)
    if r:
        print(page) 
        soup = BeautifulSoup(r.text)
        time.sleep(5)
        for link in soup.find_all(class_="structItem-title"):
            url = link.find("a").get("href")
            title = link.find("a").get_text()
            urls_dict = {"url": url, "title": title}
            app_links.append(urls_dict)
            
    else:
        print("ERROR")

https://forum.xda-developers.com/f/android-apps-and-games.530/page-2
https://forum.xda-developers.com/f/android-apps-and-games.530/page-3
https://forum.xda-developers.com/f/android-apps-and-games.530/page-4
https://forum.xda-developers.com/f/android-apps-and-games.530/page-5
https://forum.xda-developers.com/f/android-apps-and-games.530/page-6
https://forum.xda-developers.com/f/android-apps-and-games.530/page-7
https://forum.xda-developers.com/f/android-apps-and-games.530/page-8
https://forum.xda-developers.com/f/android-apps-and-games.530/page-9
https://forum.xda-developers.com/f/android-apps-and-games.530/page-10
https://forum.xda-developers.com/f/android-apps-and-games.530/page-11
https://forum.xda-developers.com/f/android-apps-and-games.530/page-12
https://forum.xda-developers.com/f/android-apps-and-games.530/page-13
https://forum.xda-developers.com/f/android-apps-and-games.530/page-14
https://forum.xda-developers.com/f/android-apps-and-games.530/page-15
https://forum.xda-developers

https://forum.xda-developers.com/f/android-apps-and-games.530/page-119
https://forum.xda-developers.com/f/android-apps-and-games.530/page-120
https://forum.xda-developers.com/f/android-apps-and-games.530/page-121
https://forum.xda-developers.com/f/android-apps-and-games.530/page-122
https://forum.xda-developers.com/f/android-apps-and-games.530/page-123
https://forum.xda-developers.com/f/android-apps-and-games.530/page-124
https://forum.xda-developers.com/f/android-apps-and-games.530/page-125
https://forum.xda-developers.com/f/android-apps-and-games.530/page-126
https://forum.xda-developers.com/f/android-apps-and-games.530/page-127
https://forum.xda-developers.com/f/android-apps-and-games.530/page-128
https://forum.xda-developers.com/f/android-apps-and-games.530/page-129
https://forum.xda-developers.com/f/android-apps-and-games.530/page-130
https://forum.xda-developers.com/f/android-apps-and-games.530/page-131
https://forum.xda-developers.com/f/android-apps-and-games.530/page-132
https:

https://forum.xda-developers.com/f/android-apps-and-games.530/page-236
https://forum.xda-developers.com/f/android-apps-and-games.530/page-237
https://forum.xda-developers.com/f/android-apps-and-games.530/page-238
https://forum.xda-developers.com/f/android-apps-and-games.530/page-239
https://forum.xda-developers.com/f/android-apps-and-games.530/page-240
https://forum.xda-developers.com/f/android-apps-and-games.530/page-241
https://forum.xda-developers.com/f/android-apps-and-games.530/page-242
https://forum.xda-developers.com/f/android-apps-and-games.530/page-243
https://forum.xda-developers.com/f/android-apps-and-games.530/page-244
https://forum.xda-developers.com/f/android-apps-and-games.530/page-245
https://forum.xda-developers.com/f/android-apps-and-games.530/page-246
https://forum.xda-developers.com/f/android-apps-and-games.530/page-247
https://forum.xda-developers.com/f/android-apps-and-games.530/page-248
https://forum.xda-developers.com/f/android-apps-and-games.530/page-249
https:

https://forum.xda-developers.com/f/android-apps-and-games.530/page-352
https://forum.xda-developers.com/f/android-apps-and-games.530/page-353
https://forum.xda-developers.com/f/android-apps-and-games.530/page-354
https://forum.xda-developers.com/f/android-apps-and-games.530/page-355
https://forum.xda-developers.com/f/android-apps-and-games.530/page-356
https://forum.xda-developers.com/f/android-apps-and-games.530/page-357
https://forum.xda-developers.com/f/android-apps-and-games.530/page-358
https://forum.xda-developers.com/f/android-apps-and-games.530/page-359
https://forum.xda-developers.com/f/android-apps-and-games.530/page-360
https://forum.xda-developers.com/f/android-apps-and-games.530/page-361
https://forum.xda-developers.com/f/android-apps-and-games.530/page-362
https://forum.xda-developers.com/f/android-apps-and-games.530/page-363
https://forum.xda-developers.com/f/android-apps-and-games.530/page-364
https://forum.xda-developers.com/f/android-apps-and-games.530/page-365
https:

https://forum.xda-developers.com/f/android-apps-and-games.530/page-468
https://forum.xda-developers.com/f/android-apps-and-games.530/page-469
https://forum.xda-developers.com/f/android-apps-and-games.530/page-470
https://forum.xda-developers.com/f/android-apps-and-games.530/page-471
https://forum.xda-developers.com/f/android-apps-and-games.530/page-472
https://forum.xda-developers.com/f/android-apps-and-games.530/page-473
https://forum.xda-developers.com/f/android-apps-and-games.530/page-474
https://forum.xda-developers.com/f/android-apps-and-games.530/page-475
https://forum.xda-developers.com/f/android-apps-and-games.530/page-476
https://forum.xda-developers.com/f/android-apps-and-games.530/page-477
https://forum.xda-developers.com/f/android-apps-and-games.530/page-478
https://forum.xda-developers.com/f/android-apps-and-games.530/page-479
https://forum.xda-developers.com/f/android-apps-and-games.530/page-480
https://forum.xda-developers.com/f/android-apps-and-games.530/page-481
https:

https://forum.xda-developers.com/f/android-apps-and-games.530/page-584
https://forum.xda-developers.com/f/android-apps-and-games.530/page-585
https://forum.xda-developers.com/f/android-apps-and-games.530/page-586
https://forum.xda-developers.com/f/android-apps-and-games.530/page-587
https://forum.xda-developers.com/f/android-apps-and-games.530/page-588
https://forum.xda-developers.com/f/android-apps-and-games.530/page-589
https://forum.xda-developers.com/f/android-apps-and-games.530/page-590
https://forum.xda-developers.com/f/android-apps-and-games.530/page-591
https://forum.xda-developers.com/f/android-apps-and-games.530/page-592
https://forum.xda-developers.com/f/android-apps-and-games.530/page-593
https://forum.xda-developers.com/f/android-apps-and-games.530/page-594
https://forum.xda-developers.com/f/android-apps-and-games.530/page-595
https://forum.xda-developers.com/f/android-apps-and-games.530/page-596
https://forum.xda-developers.com/f/android-apps-and-games.530/page-597
https:

https://forum.xda-developers.com/f/android-apps-and-games.530/page-700
https://forum.xda-developers.com/f/android-apps-and-games.530/page-701
https://forum.xda-developers.com/f/android-apps-and-games.530/page-702
https://forum.xda-developers.com/f/android-apps-and-games.530/page-703
https://forum.xda-developers.com/f/android-apps-and-games.530/page-704
https://forum.xda-developers.com/f/android-apps-and-games.530/page-705
https://forum.xda-developers.com/f/android-apps-and-games.530/page-706
https://forum.xda-developers.com/f/android-apps-and-games.530/page-707
https://forum.xda-developers.com/f/android-apps-and-games.530/page-708
https://forum.xda-developers.com/f/android-apps-and-games.530/page-709
https://forum.xda-developers.com/f/android-apps-and-games.530/page-710
https://forum.xda-developers.com/f/android-apps-and-games.530/page-711
https://forum.xda-developers.com/f/android-apps-and-games.530/page-712
https://forum.xda-developers.com/f/android-apps-and-games.530/page-713
https:

https://forum.xda-developers.com/f/android-apps-and-games.530/page-816
https://forum.xda-developers.com/f/android-apps-and-games.530/page-817
https://forum.xda-developers.com/f/android-apps-and-games.530/page-818
https://forum.xda-developers.com/f/android-apps-and-games.530/page-819
https://forum.xda-developers.com/f/android-apps-and-games.530/page-820
https://forum.xda-developers.com/f/android-apps-and-games.530/page-821
https://forum.xda-developers.com/f/android-apps-and-games.530/page-822
https://forum.xda-developers.com/f/android-apps-and-games.530/page-823
https://forum.xda-developers.com/f/android-apps-and-games.530/page-824
https://forum.xda-developers.com/f/android-apps-and-games.530/page-825
https://forum.xda-developers.com/f/android-apps-and-games.530/page-826
https://forum.xda-developers.com/f/android-apps-and-games.530/page-827
https://forum.xda-developers.com/f/android-apps-and-games.530/page-828
https://forum.xda-developers.com/f/android-apps-and-games.530/page-829
https:

https://forum.xda-developers.com/f/android-apps-and-games.530/page-932
https://forum.xda-developers.com/f/android-apps-and-games.530/page-933
https://forum.xda-developers.com/f/android-apps-and-games.530/page-934
https://forum.xda-developers.com/f/android-apps-and-games.530/page-935
https://forum.xda-developers.com/f/android-apps-and-games.530/page-936
https://forum.xda-developers.com/f/android-apps-and-games.530/page-937
https://forum.xda-developers.com/f/android-apps-and-games.530/page-938
https://forum.xda-developers.com/f/android-apps-and-games.530/page-939
https://forum.xda-developers.com/f/android-apps-and-games.530/page-940
https://forum.xda-developers.com/f/android-apps-and-games.530/page-941
https://forum.xda-developers.com/f/android-apps-and-games.530/page-942
https://forum.xda-developers.com/f/android-apps-and-games.530/page-943
https://forum.xda-developers.com/f/android-apps-and-games.530/page-944
https://forum.xda-developers.com/f/android-apps-and-games.530/page-945
https:

https://forum.xda-developers.com/f/android-apps-and-games.530/page-1047
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1048
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1049
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1050
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1051
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1052
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1053
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1054
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1055
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1056
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1057
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1058
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1059
https://forum.xda-developers.com/f/android-apps-and-games.530/pa

https://forum.xda-developers.com/f/android-apps-and-games.530/page-1161
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1162
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1163
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1164
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1165
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1166
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1167
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1168
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1169
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1170
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1171
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1172
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1173
https://forum.xda-developers.com/f/android-apps-and-games.530/pa

https://forum.xda-developers.com/f/android-apps-and-games.530/page-1275
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1276
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1277
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1278
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1279
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1280
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1281
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1282
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1283
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1284
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1285
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1286
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1287
https://forum.xda-developers.com/f/android-apps-and-games.530/pa

https://forum.xda-developers.com/f/android-apps-and-games.530/page-1574
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1575
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1576
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1577
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1578
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1579
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1580
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1581
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1582
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1583
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1584
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1585
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1586
https://forum.xda-developers.com/f/android-apps-and-games.530/pa

https://forum.xda-developers.com/f/android-apps-and-games.530/page-1688
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1689
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1690
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1691
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1692
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1693
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1694
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1695
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1696
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1697
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1698
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1699
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1700
https://forum.xda-developers.com/f/android-apps-and-games.530/pa

https://forum.xda-developers.com/f/android-apps-and-games.530/page-1802
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1803
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1804
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1805
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1806
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1807
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1808
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1809
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1810
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1811
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1812
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1813
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1814
https://forum.xda-developers.com/f/android-apps-and-games.530/pa

https://forum.xda-developers.com/f/android-apps-and-games.530/page-1916
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1917
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1918
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1919
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1920
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1921
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1922
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1923
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1924
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1925
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1926
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1927
https://forum.xda-developers.com/f/android-apps-and-games.530/page-1928
https://forum.xda-developers.com/f/android-apps-and-games.530/pa

https://forum.xda-developers.com/f/android-apps-and-games.530/page-2030
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2031
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2032
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2033
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2034
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2035
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2036
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2037
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2038
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2039
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2040
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2041
https://forum.xda-developers.com/f/android-apps-and-games.530/page-2042
https://forum.xda-developers.com/f/android-apps-and-games.530/pa

In [8]:
app_links

[{'url': '/t/app-6-0-compass-of-chance.4440211/',
  'title': '[APP][6.0+]Compass of Chance'},
 {'url': '/t/app-5-0-fairemail-fully-featured-open-source-privacy-oriented-email-app.3824168/',
  'title': '[APP][5.0+] FairEmail - Fully featured, open source, privacy oriented email app'},
 {'url': '/t/app-all-root-solutions-6-0-viper4android-fx-2-7.3774651/',
  'title': '[APP][ALL ROOT SOLUTIONS][6.0+]ViPER4Android FX 2.7'},
 {'url': '/t/xfinity-tv-mod-android-box-hdmi-output.3535881/',
  'title': 'Xfinity TV Mod Android Box (HDMI output)'},
 {'url': '/t/app-6-0-netguard-no-root-firewall.3233012/',
  'title': '[APP][6.0+] NetGuard - No-root firewall'},
 {'url': '/t/app-4-1-kiwi-browser-chromium-adblock-caf-night-mode-2021-07-29-active.3797252/',
  'title': '[APP][4.1+] Kiwi Browser? Chromium + Adblock + CAF + Night Mode 2021-07-29|Active'},
 {'url': '/t/guide-multilingo-keyboard.4064431/',
  'title': '[GUIDE] MultiLingO Keyboard'},
 {'url': '/t/app-2-2-root-wifi-reaver-gui-for-android.24568

In [41]:
r

<Response [200]>

In [42]:
app_links

[{'/t/app-4-1-simple-mobile-tools.3828023/': '/t/app-4-1-simple-mobile-tools.3828023/',
  '[APP][4.1+]Simple Mobile Tools': '[APP][4.1+]Simple Mobile Tools'},
 {'/t/app-root-nonroot-official-adaway-v5-12-1-v6-0-0b.2190753/': '/t/app-root-nonroot-official-adaway-v5-12-1-v6-0-0b.2190753/',
  '[APP][ROOT/NONROOT][OFFICIAL] AdAway v5.12.1/ v6.0.0b': '[APP][ROOT/NONROOT][OFFICIAL] AdAway v5.12.1/ v6.0.0b'},
 {'/t/app-root-4-0-battery-charge-limit-v1-1-1.3557002/': '/t/app-root-4-0-battery-charge-limit-v1-1-1.3557002/',
  '[APP][ROOT][4.0+]Battery Charge Limit v1.1.1': '[APP][ROOT][4.0+]Battery Charge Limit v1.1.1'},
 {'/t/app-notification-toggle.1037817/': '/t/app-notification-toggle.1037817/',
  '[APP] Notification Toggle': '[APP] Notification Toggle'},
 {'/t/closed-app-6-0-netguard-no-root-firewall.3233012/': '/t/closed-app-6-0-netguard-no-root-firewall.3233012/',
  '[CLOSED][APP][6.0+] NetGuard - No-root firewall': '[CLOSED][APP][6.0+] NetGuard - No-root firewall'},
 {'/t/message-forward

In [12]:
product_pages = pd.DataFrame(app_links)

In [ ]:
product_pages

In [14]:
product_pages.to_csv("womproject/product_pages.csv",index=False, sep =";")

In [17]:
with open('product_pages.json', 'w') as fp:
    json.dump(app_links, fp)

## Scraping each app's forum

Initiating data:

In [27]:
f = open('product_pages.json')
pages = json.load(f)

In [201]:
counter = 0
#for number in left:
    #page = pages[number]
for page in pages:    
    url = "https://forum.xda-developers.com" + page["url"]
    r = requests.get(url)
    if r:
        soup = BeautifulSoup(r.text)
        page["launch_date"] = soup.find('time').get_text()
        page["launch_message"] = soup.find('article', class_="message-body js-selectToQuote").get_text()
        external_links = []
        for ex_link  in soup.find('article', class_="message-body js-selectToQuote").find_all('a'):
            http = str(ex_link.get("href"))
            external_links.append(http)
        page["external_links"] = external_links
        if re.search("Deleted member", str(soup.find(class_="message-userDetails"))) is None:
            if soup.find(class_="username--style1") is None:
                page["main_developer_username"] = soup.find(class_="message-userDetails").a.get_text()
                page["main_developer_page"] = soup.find(class_="message-userDetails").a.get("href")
                page["main_developer_title"] = soup.find(class_="message-userDetails").find('h5').get_text()
                page["main_developer_joined"] = soup.find(class_="message-userExtras").find_all("dd")[0].get_text()
                page["main_developer_messages"] = soup.find(class_="message-userExtras").find_all("dd")[1].get_text()
                page["main_developer_reactions"] = soup.find(class_="message-userExtras").find_all("dd")[2].get_text()
            else: 
                page["main_developer_username"] = soup.find(class_="username--style1").get_text()
                page["main_developer_title"] = soup.find(class_="message-userDetails").find('h5').get_text()
        
            if soup.find(class_="blockStatus") is not None:
                page["block_message"] ="yes"
        else: 
            page["main_developer_username"] = "Deleted member"
        time.sleep(3)
        counter += 1
        if counter % 100 == 0:
            print(counter)

100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
5300
5400
5500
5600
5700
5800
5900
6000
6100
6200
6300
6400
6500
6600
6700
6800
6900
7000
7100
7200
7300
7400
7500
7600
7700
7800
7900
8000
8100
8200
8300
8400
8500
8600
8700
8800
8900
9000
9100
9200
9300
9400
9500
9600
9700
9800
9900
10000
10100
10200
10300
10400
10500
10600
10700
10800
10900
11000
11100
11200
11300
11400
11500
11600
11700
11800
11900
12000
12100
12200
12300


In [202]:
pages_pd = pd.DataFrame(pages)

In [194]:
#left = list(pages_pd[pd.DataFrame(pages)["launch_date"].isnull()==True].index)

In [204]:
pages_pd.to_csv("womproject/scraped_product_pages.csv",index=False, sep =";")

In [205]:
with open('womproject/data/scraped_product_pages.json', 'w') as fp:
    json.dump(pages, fp)

# Extracting useful data

Load data:

In [60]:
f = open('data/scraped_product_pages.json')
pages = json.load(f)

In [61]:
pages_pd = pd.DataFrame(pages)

In [62]:
pages_pd.loc[2, "external_links"]

['https://labs.xda-developers.com/store/app/com.pittvandewitt.viperfx',
 'https://forum.xda-developers.com/m/6037748/',
 'https://zackptg5.com/android.php#viper',
 'http://t.me/vdcservice',
 'https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=53H9TP89FLWUU',
 'https://labs.xda-developers.com/store/app/com.pittvandewitt.viperfx',
 'https://forum.xda-developers.com/member.php?u=9073985',
 'https://forum.xda-developers.com/member.php?u=4539728',
 'https://forum.xda-developers.com/member.php?u=7052175']

## Extract GitHub and Google Play links

In [63]:
pages_pd["code"] = pages_pd["title"].str.extract("(^\[.+?\])")

In [64]:
pages_pd["app"] = pages_pd["title"].str.extract("([Aa][Pp]{2})")

In [65]:
pages_pd[pages_pd["app"].isnull()==False]

url  \
0                  /t/app-6-0-compass-of-chance.4440211/   
1      /t/app-5-0-fairemail-fully-featured-open-sourc...   
2      /t/app-all-root-solutions-6-0-viper4android-fx...   
4          /t/app-6-0-netguard-no-root-firewall.3233012/   
5      /t/app-4-1-kiwi-browser-chromium-adblock-caf-n...   
...                                                  ...   
38430            /t/essential-apps-for-new-users.673395/   
38432     /t/30-000-apps-and-no-bridge-card-game.672118/   
38435                  /t/market-as-separate-app.672499/   
38437           /t/1-50-bank-charge-on-every-app.672213/   
38438         /t/request-relationship-helper-app.669411/   

                                                   title   launch_date  \
0                           [APP][6.0+]Compass of Chance   May 2, 2022   
1      [APP][5.0+] FairEmail - Fully featured, open s...   Aug 2, 2018   
2      [APP][ALL ROOT SOLUTIONS][6.0+]ViPER4Android F...   Apr 7, 2018   
4                [APP][6.0+] NetGuard - No-root firewall  Oct 25, 2015   
5      [APP][4.1+] Kiwi Browser? Chromium + Adblock +...  May 30, 2018   
...                                                  ...           ...   
38430                      Essential Apps for new users?  Apr 30, 2010   
38432              30,000+ apps and no Bridge Card game?  Apr 28, 2010   
38435                             Market as separate app  Apr 29, 2010   
38437                    £1.50 bank charge on every app.  Apr 28, 2010   
38438                  [REQUEST] Relationship helper app  Apr 23, 2010   

                                          launch_message  \
0      \nThe compass does not point to the north, but...   
1      \n\n\n\n\n\nFairEmail\nOpen source, privacy fr...   
2      \nViPER4Android FX\n\n\n\n\n\nBy pittvandewitt...   
4      \n\n\n\n\n\n\nNetGuard provides simple and adv...   
5      \nHi,\n\n*** Join the Kiwi development Discord...   
...                                                  ...   
38430  \nThis thread used to have a sticky that had a...   
38432  \nOr can someone please point me in the right ...   
38435  \nHey people, I rebooted my phone this morning...   
38437  \nNot seen too many people moan about this, bu...   
38438  \nI need an app that will take care of the tex...   

                                          external_links  \
0      [https://play.google.com/store/apps/details?id...   
1      [https://github.com/M66B/open-source-email/, h...   
2      [https://labs.xda-developers.com/store/app/com...   
4      [https://contact.faircode.eu/?product=netguard...   
5      [https://discord.gg/XyMppQq, http://www.coinhi...   
...                                                  ...   
38430                                                 []   
38432                                                 []   
38435                                                 []   
38437                                                 []   
38438                                                 []   

      main_developer_username        main_developer_page  \
0                     Baiumka       /m/baiumka.12120441/   
1                        M66B           /m/m66b.2799345/   
2                 Team_DeWitt    /m/team_dewitt.9073985/   
4                        M66B           /m/m66b.2799345/   
5                    arnaud42       /m/arnaud42.9196003/   
...                       ...                        ...   
38430               itpromike      /m/itpromike.2067307/   
38432           RaymondScrott  /m/raymondscrott.1400603/   
38435                 fuz1987        /m/fuz1987.2336665/   
38437              Phil750123     /m/phil750123.2481769/   
38438                 AdrianK                        NaN   

       main_developer_title main_developer_joined main_developer_messages  \
0                New member           May 2, 2022                       4   
1      Recognized Developer           Aug 1, 2010                  25,652   
2             Senior Member           Apr 7, 2018

In [66]:
pd.options.display.max_rows = 999
pages_pd["code"].value_counts()

[APP]                     9440
[GAME]                    5070
[Q]                       1631
[App]                     1621
[Game]                    1551
                          ... 
[ONLINE GAME]                1
[SPL]                        1
[S Health v4.5.0.0113]       1
[APP WALLPAPER]              1
[ Tool / App ]               1
Name: code, Length: 1391, dtype: int64

In [32]:
df_external_links = pages_pd["external_links"].copy()

In [33]:
df_expand_links = df_external_links.apply(pd.Series)

In [34]:
df_expand_links

0    \
0      https://play.google.com/store/apps/details?id=...   
1             https://github.com/M66B/open-source-email/   
2      https://labs.xda-developers.com/store/app/com....   
3                                                    NaN   
4       https://contact.faircode.eu/?product=netguard%2B   
...                                                  ...   
38435                                                NaN   
38436                                                NaN   
38437                                                NaN   
38438                                                NaN   
38439                                                NaN   

                                                     1    \
0      https://forum.xda-developers.com/attachments/l...   
1      https://github.com/M66B/open-source-email#user...   
2            https://forum.xda-developers.com/m/6037748/   
3                                                    NaN   
4      https://github.com/M66B/NetGuard/blob/master/A...   
...                                                  ...   
38435                                                NaN   
38436                                                NaN   
38437                                                NaN   
38438                                                NaN   
38439                                                NaN   

                                                     2    \
0                                                    NaN   
1      https://www.xda-developers.com/fairemail-open-...   
2                 https://zackptg5.com/android.php#viper   
3                                                    NaN   
4              https://github.com/M66B/NetGuard#netguard   
...                                                  ...   
38435                                                NaN   
38436                                                NaN   
38437                                                NaN   
38438                                                NaN   
38439                                                NaN   

                                                  3    \
0                                                 NaN   
1              https://email.faircode.eu/#screenshots   
2                              http://t.me/vdcservice   
3                                                 NaN   
4      https://github.com/M66B/NetGuard#compatibility   
...                                               ...   
38435                                             NaN   
38436                                             NaN   
38437                                             NaN   
38438                                             NaN   
38439                                             NaN   

                                                     4    \
0                                                    NaN   
1      https://github.com/M66B/open-source-email/blob...   
2      https://www.paypal.com/cgi-bin/webscr?cmd=_s-x...   
3                                                    NaN   
4      https://github.com/M66B/NetGuard/blob/master/F...   
...                                                  ...   
38435                                                NaN   
38436                                                NaN   
38437                                                NaN   
38438                                                NaN   
38439                                                NaN   

                                                     5    \
0                                                    NaN   
1      https://github.com/M66B/open-source-email/blob...   
2      https://labs.xda-developers.com/store/app/com....   
3                                                    NaN   
4          https://github.com/M66B/NetGuard/#permissions   
...                                                  ...   
38435                                                NaN   
3843

In [51]:
#df_expand_links[df_expand_links[191].isnull()==False]

In [35]:
df_expand_links = df_expand_links.replace(np.nan,'',regex=True)

In [23]:
df_expand_links

0    \
0      https://play.google.com/store/apps/details?id=...   
1             https://github.com/M66B/open-source-email/   
2      https://labs.xda-developers.com/store/app/com....   
3                                                          
4       https://contact.faircode.eu/?product=netguard%2B   
...                                                  ...   
38435                                                      
38436                                                      
38437                                                      
38438                                                      
38439                                                      

                                                     1    \
0      https://forum.xda-developers.com/attachments/l...   
1      https://github.com/M66B/open-source-email#user...   
2            https://forum.xda-developers.com/m/6037748/   
3                                                          
4      https://github.com/M66B/NetGuard/blob/master/A...   
...                                                  ...   
38435                                                      
38436                                                      
38437                                                      
38438                                                      
38439                                                      

                                                     2    \
0                                                          
1      https://www.xda-developers.com/fairemail-open-...   
2                 https://zackptg5.com/android.php#viper   
3                                                          
4              https://github.com/M66B/NetGuard#netguard   
...                                                  ...   
38435                                                      
38436                                                      
38437                                                      
38438                                                      
38439                                                      

                                                  3    \
0                                                       
1              https://email.faircode.eu/#screenshots   
2                              http://t.me/vdcservice   
3                                                       
4      https://github.com/M66B/NetGuard#compatibility   
...                                               ...   
38435                                                   
38436                                                   
38437                                                   
38438                                                   
38439                                                   

                                                     4    \
0                                                          
1      https://github.com/M66B/open-source-email/blob...   
2      https://www.paypal.com/cgi-bin/webscr?cmd=_s-x...   
3                                                          
4      https://github.com/M66B/NetGuard/blob/master/F...   
...                                                  ...   
38435                                                      
38436                                                      
38437                                                      
38438                                                      
38439                                                      

                                                     5    \
0                                                          
1      https://github.com/M66B/open-source-email/blob...   
2      https://labs.xda-developers.com/store/app/com....   
3                                                          
4          https://github.com/M66B/NetGuard/#permissions   
...                                                  ...   
38435                                                      
3843

In [67]:
pages_pd['githublink']=df_expand_links.apply(','.join,axis=1).str.extract(r'(https\:\/\/github\.com\/[^\/\']+?\/[^\/\']+?)[\/,]', expand = False)

In [68]:
pages_pd['github_dev']=df_expand_links.apply(','.join,axis=1).str.extract(r'(https\:\/\/github\.com\/.+?)[\/\']', expand = False)

In [69]:
pages_pd[pages_pd["github_dev"].isnull()==False]

url  \
1      /t/app-5-0-fairemail-fully-featured-open-sourc...   
4          /t/app-6-0-netguard-no-root-firewall.3233012/   
9      /t/app-root-nonroot-official-adaway-v5-12-1-v6...   
10     /t/app-magisk-module-acc-custom-accent-colour-...   
12     /t/app-root-4-0-battery-charge-limit-v1-1-1.35...   
19               /t/app-4-1-simple-mobile-tools.3828023/   
20               /t/app-4-1-simple-mobile-tools.3828023/   
25     /t/app-4-4-amaze-file-utilities-analyse-transf...   
35     /t/app-fix-100-working-viper4android-for-andro...   
38     /t/app-root-5-0-vpn-hotspot-vpn-tethering-wi-f...   
40     /t/app-4-0-3-4-1-rotation-orientation-manager-...   
48     /t/app-4-1-ymusic-listen-youtube-music-in-back...   
49     /t/app-root-11-23-03-classic-power-menu-power-...   
51     /t/app-android-5-0-3-4-120-21-05-2020-adguard-...   
54     /t/app-patch-enable-messenger-original-chat-he...   
58        /t/app-android-tv-projectivy-launcher.4436549/   
64     /t/game-2-3-retroarch-android-emulator-v1-0-0-...   
65     /t/app-4-4-bromite-chromium-plus-adblocking.36...   
67                    /t/app-mod-pixel-launcher.4442077/   
68     /t/app-7-0-whatsrunning-know-whats-running-on-...   
69                    /t/app-6-0-systemui-tuner.3588675/   
73               /t/tool-whatsapp-viewer-for-pc.2719741/   
76               /t/app-mod-root-nonroot-vanced.3758757/   
79     /t/app-4-0-keepassdx-open-source-password-mana...   
80     /t/tool-whatsapp-google-drive-extractor-update...   
86     /t/5-0-root-microtoggles-an-opensource-toggle-...   
88     /t/app-root-nonroot-smarthertz-control-your-di...   
102               /t/app-5-0-intercepter-ng-2-8.2030865/   
104    /t/app-4-2-secondscreen-better-screen-mirrorin...   
105    /t/app-root-12-11-05-pixel-launcher-mods-icon-...   
106    /t/app-port-lineageos-15-1-trebuchet-laucher-f...   
109    /t/app-root-5-0-1st-nov-2020-migrate-custom-ro...   
111    /t/app-1-3-tap-tap-double-tap-on-back-of-devic...   
121    /t/app-5-0-foss-root-adb-app-manager-most-adva...   
129    /t/app-v1-10-vd-infos-package-com-vitaodoidao-...   
133            /t/mod-app-gallery-quickpic-2022.3790425/   
137    /t/app-4-4-aurora-store-open-source-play-store...   
140    /t/app-5-0-root-fde-ai-v22-04-1-powerful-aio-a...   
151    /t/app-5-0-root-nonroot-gplv3-material-files-o...   
165    /t/best-app-for-watching-streaming-downloading...   
170    /t/game-droid-4-0-doom3-diii4a-roe-the-lost-mi...   
173           /t/root-pornaway-block-porn-sites.3460036/   
186    /t/app-foss-gsf-widevine-l1-2-3-treble-a-b-dev...   
189        /t/app-8-0-skip-track-by-volume-keys.3914337/   
190    /t/app-microg-gmscore-lightweight-free-softwar...   
194     /t/app-4-1-unofficial-searx-app-pi-hole.4246935/   
201                   /t/app-4-0-kernel-adiutor.2986129/   
202    /t/app-4-1-barquode-matrix-manager-3-4-1-08-04...   
206    /t/app-5-1-trackercontrol-control-trackers-and...   
209    /t/app-5-0-spectrum-a-kernel-manager-for-users...   
215    /t/lawnchair-customizable-pixel-launcher.3627137/   
222    /t/app-root-6-tether-tproxy-v0-1-usb-wifi-hots...   
233    /t/app-mod-4-4-instagram-without-ads-and-annoy...   
238    /t/app-2-1-15-jun-2-1-better-wifi-on-off.1801284/   
241    /t/introduce-brevent-black-prevent-without-roo...   
244    /t/app-root-shizuku-10-11-feb-darq-per-app-sel...   
245    /t/app-mod-netflix-kodi-edition-kodi-fork-with...   
247            /t/jamesdsp-audio-manager-mmt-ex.3607970/   
255    /t/app-root-snapchat-dark-mode-enable-dark-mod...   
273    /t/port-app-5-0-06-06-21-samsung-music-port-fo...   
286    /t/app-4-0-amaze-file-manager-material-theme.2...   
288    /t/app-netflix-hd-for-widevine-l1-device.3942467/   
298            /t/black-hole-a-music-player-app.4430223/   
308    /t/app-8-0-open-source-root-oandbackupx-apps-a...   
316    /t/app-4-0-3-gplv3-apkupdater-2-get-updates-fo...   
317    /t/app-tool-2-0-official-the-selinux-switch.36...   
318    /t/app-mod-fiil-chinese-ap

In [70]:
blocked = pages_pd[pages_pd["block_message"].isnull()==False].index

In [71]:
pages_pd = pages_pd.drop(blocked)

In [72]:
pages_pd.loc[12,"githublink"]

'https://github.com/sriharshaarangi/BatteryChargeLimit'

In [73]:
pages_pd['googleplay_link']=df_expand_links.apply(','.join,axis=1).str.extract(r'(https\:\/\/play\.google\.com\/store\/apps\/details\?id\=.+?),', expand = 
False)

In [74]:
len(pages_pd[pages_pd["githublink"].isnull()==False].index)

956

App pages with GitHub link available: 956??

This is the starting sample

In [75]:
len(pages_pd[(pages_pd["app"].isnull()==False)&(pages_pd['githublink'].isnull()==False)].index)

777

In [76]:
len(pages_pd[(pages_pd["app"].isnull()==False)&(pages_pd['githublink'].isnull()==False) & (pages_pd['googleplay_link'].isnull()==False)].index)

394

## Extract contributors links:

In [77]:
pages_pd["other_contributors"] = pages_pd["launch_message"].str.contains("[Cc]ontributors|[Tt]eam", regex=True)

In [78]:
pages_pd.loc[2, "external_links"]

['https://labs.xda-developers.com/store/app/com.pittvandewitt.viperfx',
 'https://forum.xda-developers.com/m/6037748/',
 'https://zackptg5.com/android.php#viper',
 'http://t.me/vdcservice',
 'https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=53H9TP89FLWUU',
 'https://labs.xda-developers.com/store/app/com.pittvandewitt.viperfx',
 'https://forum.xda-developers.com/member.php?u=9073985',
 'https://forum.xda-developers.com/member.php?u=4539728',
 'https://forum.xda-developers.com/member.php?u=7052175']

In [79]:
contributors = df_expand_links.apply(','.join,axis=1).str.extractall(r'(https\:\/\/forum\.xda\-developers\.com\/member\.php\?u.+?),').unstack()         
contributors.columns = contributors.columns.droplevel(0)
#contributors

In [80]:
new_names = ['contrlink_1','contrlink_2','contrlink_3','contrlink_4', 'contrlink_5','contrlink_6', 'contrlink_7']
contributors.rename(columns=dict(zip(list(contributors.columns), new_names)), inplace=True)

In [81]:
contributors.count(axis=1)

1        1
2        3
4        1
9        2
10       3
12       2
46       1
51       3
62       1
66       1
70       1
86       1
103      1
104      1
112      1
118      1
123      1
129      7
137      1
151      1
156      1
180      3
181      1
190      1
195      1
200      1
206      1
209      1
215      3
225      1
247      3
248      2
252      1
287      1
316      1
317      1
364      1
378      1
379      2
380      1
382      1
398      1
405      1
410      1
416      1
431      1
439      1
441      1
443      1
463      2
473      1
474      1
475      2
509      1
545      3
549      1
552      1
570      1
581      1
590      1
591      1
593      1
595      1
606      1
638      1
673      2
753      1
759      1
783      1
806      1
824      1
827      1
874      2
885      1
886      1
901      1
912      1
929      1
938      1
941      1
957      1
978      1
992      1
1030     2
1049     1
1080     1
1096     2
1099     1
1108     1
1110     1
1121     1

In [82]:
contributors["number_of_developers"] = contributors.count(axis=1)

In [83]:
contributors

match                                        contrlink_1  \
1      https://forum.xda-developers.com/member.php?u=...   
2      https://forum.xda-developers.com/member.php?u=...   
4      https://forum.xda-developers.com/member.php?u=...   
9      https://forum.xda-developers.com/member.php?u=...   
10     https://forum.xda-developers.com/member.php?u=...   
12     https://forum.xda-developers.com/member.php?u=...   
46     https://forum.xda-developers.com/member.php?u=...   
51     https://forum.xda-developers.com/member.php?u=...   
62     https://forum.xda-developers.com/member.php?u=...   
66     https://forum.xda-developers.com/member.php?u=...   
70     https://forum.xda-developers.com/member.php?u=...   
86     https://forum.xda-developers.com/member.php?u=...   
103    https://forum.xda-developers.com/member.php?u=...   
104    https://forum.xda-developers.com/member.php?u=...   
112    https://forum.xda-developers.com/member.php?u=...   
118    https://forum.xda-developers.com/member.php?u=...   
123    https://forum.xda-developers.com/member.php?u=...   
129    https://forum.xda-developers.com/member.php?u=...   
137    https://forum.xda-developers.com/member.php?u=...   
151    https://forum.xda-developers.com/member.php?u=...   
156    https://forum.xda-developers.com/member.php?u=...   
180    https://forum.xda-developers.com/member.php?u=...   
181    https://forum.xda-developers.com/member.php?u=...   
190    https://forum.xda-developers.com/member.php?u=...   
195    https://forum.xda-developers.com/member.php?u=...   
200    https://forum.xda-developers.com/member.php?u=...   
206    https://forum.xda-developers.com/member.php?u=...   
209    https://forum.xda-developers.com/member.php?u=...   
215    https://forum.xda-developers.com/member.php?u=...   
225    https://forum.xda-developers.com/member.php?u=...   
247    https://forum.xda-developers.com/member.php?u=...   
248    https://forum.xda-developers.com/member.php?u=...   
252    https://forum.xda-developers.com/member.php?u=...   
287    https://forum.xda-developers.com/member.php?u=...   
316    https://forum.xda-developers.com/member.php?u=...   
317    https://forum.xda-developers.com/member.php?u=...   
364    https://forum.xda-developers.com/member.php?u=...   
378    https://forum.xda-developers.com/member.php?u=...   
379    https://forum.xda-developers.com/member.php?u=...   
380    https://forum.xda-developers.com/member.php?u=...   
382    https://forum.xda-developers.com/member.php?u=...   
398    https://forum.xda-developers.com/member.php?u=...   
405    https://forum.xda-developers.com/member.php?u=...   
410    https://forum.xda-developers.com/member.php?u=...   
416    https://forum.xda-developers.com/member.php?u=...   
431    https://forum.xda-developers.com/member.php?u=...   
439    https://forum.xda-developers.com/member.php?u=...   
441    https://forum.xda-developers.com/member.php?u=...   
443    https://forum.xda-developers.com/member.php?u=...   
463    https://forum.xda-developers.com/member.php?u=...   
473    https://forum.xda-developers.com/member.php?u=...   
474    https://forum.xda-developers.com/member.php?u=...   
475    https://forum.xda-developers.com/member.php?u=...   
509    https://forum.xda-developers.com/member.php?u=...   
545    https://forum.xda-developers.com/member.php?u=...   
549    https://forum.xda-developers.com/member.php?u=...   
552    https://forum.xda-developers.com/member.php?u=...   
570    https://forum.xda-developers.com/member.php?u=...   
581    https://forum.xda-developers.com/member.php?u=...   
590    https://forum.xda-developers.com/member.php?u=...   
591    https://forum.xda-developers.com/member.php?u=...   
593    https://forum.xda-developers.com/member.php?u=...   
595    https://forum.xda-developers.com/member.php?u=...   
606    https://forum.xda-developers.com/member.php?u=...   
638    https://forum.xda-developers.com/member.php?u=...   
673    https://forum.xda-developers.com/

In [84]:
pages_pd = pages_pd.join(contributors)

In [85]:
pages_pd['number_of_developers'] = pages_pd['number_of_developers'].fillna(1)

In [86]:
pages_pd

url  \
0                  /t/app-6-0-compass-of-chance.4440211/   
1      /t/app-5-0-fairemail-fully-featured-open-sourc...   
2      /t/app-all-root-solutions-6-0-viper4android-fx...   
3      /t/xfinity-tv-mod-android-box-hdmi-output.3535...   
4          /t/app-6-0-netguard-no-root-firewall.3233012/   
...                                                  ...   
38435                  /t/market-as-separate-app.672499/   
38436      /t/request-side-scrolling-online-game.672503/   
38437           /t/1-50-bank-charge-on-every-app.672213/   
38438         /t/request-relationship-helper-app.669411/   
38439                    /t/fring-some-questions.672240/   

                                                   title   launch_date  \
0                           [APP][6.0+]Compass of Chance   May 2, 2022   
1      [APP][5.0+] FairEmail - Fully featured, open s...   Aug 2, 2018   
2      [APP][ALL ROOT SOLUTIONS][6.0+]ViPER4Android F...   Apr 7, 2018   
3               Xfinity TV Mod Android Box (HDMI output)  Jan 11, 2017   
4                [APP][6.0+] NetGuard - No-root firewall  Oct 25, 2015   
...                                                  ...           ...   
38435                             Market as separate app  Apr 29, 2010   
38436               (REQUEST) Side Scrolling online game  Apr 29, 2010   
38437                    £1.50 bank charge on every app.  Apr 28, 2010   
38438                  [REQUEST] Relationship helper app  Apr 23, 2010   
38439                             Fring - Some questions  Apr 28, 2010   

                                          launch_message  \
0      \nThe compass does not point to the north, but...   
1      \n\n\n\n\n\nFairEmail\nOpen source, privacy fr...   
2      \nViPER4Android FX\n\n\n\n\n\nBy pittvandewitt...   
3      \nIm modded an apk for Xfinity TV (new and lat...   
4      \n\n\n\n\n\n\nNetGuard provides simple and adv...   
...                                                  ...   
38435  \nHey people, I rebooted my phone this morning...   
38436  \nIs it possible to create something along the...   
38437  \nNot seen too many people moan about this, bu...   
38438  \nI need an app that will take care of the tex...   
38439  \nI don't seem to be able to see a way to chan...   

                                          external_links  \
0      [https://play.google.com/store/apps/details?id...   
1      [https://github.com/M66B/open-source-email/, h...   
2      [https://labs.xda-developers.com/store/app/com...   
3                                                     []   
4      [https://contact.faircode.eu/?product=netguard...   
...                                                  ...   
38435                                                 []   
38436                                                 []   
38437                                                 []   
38438                                                 []   
38439                                                 []   

      main_developer_username        main_developer_page  \
0                     Baiumka       /m/baiumka.12120441/   
1                        M66B           /m/m66b.2799345/   
2                 Team_DeWitt    /m/team_dewitt.9073985/   
3                aiemassfiria   /m/aiemassfiria.7985572/   
4                        M66B           /m/m66b.2799345/   
...                       ...                        ...   
38435                 fuz1987        /m/fuz1987.2336665/   
38436              fadentales     /m/fadentales.2219089/   
38437              Phil750123     /m/phil750123.2481769/   
38438                 AdrianK                        NaN   
38439           TheOriginalKi  /m/theoriginalki.2518563/   

       main_developer_title main_developer_joined main_developer_messages  \
0                New member           May 2, 2022                       4   
1      Recognized Developer           Aug 1, 2010                  25,652   
2             Senior Member           Apr 7, 2018

Additional check of the other_contributors is needed: it may be that the page lists only the main developer or onlle the helping developer (compare link to the profile to the main developer_link)

## Non-Zero Price

In [87]:
pages_pd["nonzero_price"] = pages_pd["launch_message"].str.contains("purchase|paid|price|donate|donation", regex=True)

In [89]:
len(pages_pd[(pages_pd["nonzero_price"]==True)&(pages_pd['githublink'].isnull()==False)].index)

124

In [90]:
pages_pd

url  \
0                  /t/app-6-0-compass-of-chance.4440211/   
1      /t/app-5-0-fairemail-fully-featured-open-sourc...   
2      /t/app-all-root-solutions-6-0-viper4android-fx...   
3      /t/xfinity-tv-mod-android-box-hdmi-output.3535...   
4          /t/app-6-0-netguard-no-root-firewall.3233012/   
...                                                  ...   
38435                  /t/market-as-separate-app.672499/   
38436      /t/request-side-scrolling-online-game.672503/   
38437           /t/1-50-bank-charge-on-every-app.672213/   
38438         /t/request-relationship-helper-app.669411/   
38439                    /t/fring-some-questions.672240/   

                                                   title   launch_date  \
0                           [APP][6.0+]Compass of Chance   May 2, 2022   
1      [APP][5.0+] FairEmail - Fully featured, open s...   Aug 2, 2018   
2      [APP][ALL ROOT SOLUTIONS][6.0+]ViPER4Android F...   Apr 7, 2018   
3               Xfinity TV Mod Android Box (HDMI output)  Jan 11, 2017   
4                [APP][6.0+] NetGuard - No-root firewall  Oct 25, 2015   
...                                                  ...           ...   
38435                             Market as separate app  Apr 29, 2010   
38436               (REQUEST) Side Scrolling online game  Apr 29, 2010   
38437                    £1.50 bank charge on every app.  Apr 28, 2010   
38438                  [REQUEST] Relationship helper app  Apr 23, 2010   
38439                             Fring - Some questions  Apr 28, 2010   

                                          launch_message  \
0      \nThe compass does not point to the north, but...   
1      \n\n\n\n\n\nFairEmail\nOpen source, privacy fr...   
2      \nViPER4Android FX\n\n\n\n\n\nBy pittvandewitt...   
3      \nIm modded an apk for Xfinity TV (new and lat...   
4      \n\n\n\n\n\n\nNetGuard provides simple and adv...   
...                                                  ...   
38435  \nHey people, I rebooted my phone this morning...   
38436  \nIs it possible to create something along the...   
38437  \nNot seen too many people moan about this, bu...   
38438  \nI need an app that will take care of the tex...   
38439  \nI don't seem to be able to see a way to chan...   

                                          external_links  \
0      [https://play.google.com/store/apps/details?id...   
1      [https://github.com/M66B/open-source-email/, h...   
2      [https://labs.xda-developers.com/store/app/com...   
3                                                     []   
4      [https://contact.faircode.eu/?product=netguard...   
...                                                  ...   
38435                                                 []   
38436                                                 []   
38437                                                 []   
38438                                                 []   
38439                                                 []   

      main_developer_username        main_developer_page  \
0                     Baiumka       /m/baiumka.12120441/   
1                        M66B           /m/m66b.2799345/   
2                 Team_DeWitt    /m/team_dewitt.9073985/   
3                aiemassfiria   /m/aiemassfiria.7985572/   
4                        M66B           /m/m66b.2799345/   
...                       ...                        ...   
38435                 fuz1987        /m/fuz1987.2336665/   
38436              fadentales     /m/fadentales.2219089/   
38437              Phil750123     /m/phil750123.2481769/   
38438                 AdrianK                        NaN   
38439           TheOriginalKi  /m/theoriginalki.2518563/   

       main_developer_title main_developer_joined main_developer_messages  \
0                New member           May 2, 2022                       4   
1      Recognized Developer           Aug 1, 2010                  25,652   
2             Senior Member           Apr 7, 2018

In [121]:
ind_nan = pages_pd[pages_pd["nonzero_price"].isnull()].index

In [122]:
pages_pd = pages_pd.drop(ind_nan)

In [91]:
pages_pd.to_csv('scraped_product_pages_extracted.csv', sep=";")  

# Filter out product pages

In [126]:
pages_pd.loc[pages_pd["launch_message"].str.contains("http", regex = True), "ext_links_ind"] = True

In [128]:
pages_pd["ext_links_ind"].fillna(False)

0        False
1         True
2        False
3        False
4         True
         ...  
38435    False
38436    False
38437    False
38438    False
38439    False
Name: ext_links_ind, Length: 37878, dtype: bool

In [129]:
pages_clean = pages_pd[pages_pd["ext_links_ind"]==True]

In [131]:
pages_clean = pages_clean[(pages_clean["app"].isnull()==False) & (pages_clean["launch_message"].str.contains("recommend|top|overview|digest|released")==False)]

In [133]:
pages_clean.to_csv("data/safe_obs_dataset.csv", sep=";")

# Scraping available GitHub pages

closed issues/all issues; number of downloads if available

In [178]:
#setting up selenium to function
edgeOption = webdriver.EdgeOptions()
edgeOption.use_chromium = True
edgeOption.add_argument("start-maximized")
edgeOption.binary_location = r"C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"
s=service.Service(r'C:\Users\Yana Gerina\Documents\msedgedriver.exe')
driver = webdriver.Edge(service=s, options=edgeOption)

Using only products that have a github link:

In [229]:
pages_github = pages_pd[pages_pd["githublink"].isnull()==False][["url", "githublink"]].to_dict(orient="records")

In [230]:
pages_github

[{'url': '/t/app-5-0-fairemail-fully-featured-open-source-privacy-oriented-email-app.3824168/',
  'githublink': 'https://github.com/M66B/open-source-email'},
 {'url': '/t/app-6-0-netguard-no-root-firewall.3233012/',
  'githublink': 'https://github.com/M66B/NetGuard'},
 {'url': '/t/app-root-nonroot-official-adaway-v5-12-1-v6-0-0b.2190753/',
  'githublink': 'https://github.com/AdAway/AdAway'},
 {'url': '/t/app-magisk-module-acc-custom-accent-colour-creator-android-8.4011747/',
  'githublink': 'https://github.com/Akilesh-T/ACC'},
 {'url': '/t/app-root-4-0-battery-charge-limit-v1-1-1.3557002/',
  'githublink': 'https://github.com/sriharshaarangi/BatteryChargeLimit'},
 {'url': '/t/app-fix-100-working-viper4android-for-android-10-11-12-devices.4213647/',
  'githublink': 'https://github.com/MrBIMC/SELinuxModeChanger'},
 {'url': '/t/app-root-5-0-vpn-hotspot-vpn-tethering-wi-fi-repeater-open-source.3741604/',
  'githublink': 'https://github.com/Mygod/VPNHotspot'},
 {'url': '/t/app-root-11-23-03

Loop over all available github product pages:

In [233]:
for page in pages_github:
    driver.get(page["githublink"])
    #Collect the number of open and closed issues if it's available:
    if driver.find_elements(by=By.XPATH, value='//*[@id="issues-tab"]'):
        #Click on Issues tab:
        driver.find_element(by=By.XPATH, value='//*[@id="issues-tab"]').click()
        time.sleep(5)
        if len(driver.find_elements(by=By.XPATH, value="//a[contains(@href, '/issues?q=is%3Aopen')]"))>1:
            list_open = driver.find_elements(by=By.XPATH, value="//a[contains(@href, '/issues?q=is%3Aopen')]")
            iss_open = ','.join(s.text for s in list_open if s.text)
            list_closed = driver.find_elements(by=By.XPATH, value="//a[contains(@href, '/issues?q=is%3Aissue+is%3Aclosed')]")
            iss_closed = ','.join(s.text for s in list_closed if s.text)
            page["open_issues"]= re.search(r"[\d,]+", iss_open).group(0)
            page["closed_issues"] = re.search(r"[\d,]+", iss_closed).group(0)
    time.sleep(2)
    
driver.quit()    

In [176]:
#driver.quit()  
#driver.get(page["githublink"])

In [235]:
df_pages = pd.DataFrame(pages_github)

In [236]:
df_pages.to_csv("product_pages_with_github_issues.csv",index=False, sep =";")

# Full Dataset

Load data:

In [102]:
df_pages = pd.read_csv('product_pages_with_github_issues.csv', sep=";")  

In [103]:
pages_pd = pd.read_csv('scraped_product_pages_extracted.csv', sep=";", index_col = "Unnamed: 0") 

/home/jane/miniconda2/envs/briq/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2714: DtypeWarning: Columns (19,20,21,22,23,24,25) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [107]:
pages = pd.merge(pages_pd, df_pages, how="right", on=["url", "githublink"])

In [112]:
pages

url  \
0    /t/app-5-0-fairemail-fully-featured-open-sourc...   
1        /t/app-6-0-netguard-no-root-firewall.3233012/   
2    /t/app-root-nonroot-official-adaway-v5-12-1-v6...   
3    /t/app-magisk-module-acc-custom-accent-colour-...   
4    /t/app-root-4-0-battery-charge-limit-v1-1-1.35...   
5    /t/app-fix-100-working-viper4android-for-andro...   
6    /t/app-root-5-0-vpn-hotspot-vpn-tethering-wi-f...   
7    /t/app-root-11-23-03-classic-power-menu-power-...   
8    /t/app-android-5-0-3-4-120-21-05-2020-adguard-...   
9    /t/app-patch-enable-messenger-original-chat-he...   
10   /t/game-2-3-retroarch-android-emulator-v1-0-0-...   
11   /t/app-4-4-bromite-chromium-plus-adblocking.36...   
12             /t/tool-whatsapp-viewer-for-pc.2719741/   
13   /t/app-4-0-keepassdx-open-source-password-mana...   
14   /t/tool-whatsapp-google-drive-extractor-update...   
15   /t/5-0-root-microtoggles-an-opensource-toggle-...   
16   /t/app-root-nonroot-smarthertz-control-your-di...   
17   /t/app-root-12-11-05-pixel-launcher-mods-icon-...   
18   /t/app-port-lineageos-15-1-trebuchet-laucher-f...   
19   /t/app-1-3-tap-tap-double-tap-on-back-of-devic...   
20   /t/app-5-0-foss-root-adb-app-manager-most-adva...   
21   /t/app-v1-10-vd-infos-package-com-vitaodoidao-...   
22   /t/app-4-4-aurora-store-open-source-play-store...   
23   /t/app-5-0-root-fde-ai-v22-04-1-powerful-aio-a...   
24   /t/app-5-0-root-nonroot-gplv3-material-files-o...   
25   /t/game-droid-4-0-doom3-diii4a-roe-the-lost-mi...   
26          /t/root-pornaway-block-porn-sites.3460036/   
27   /t/app-foss-gsf-widevine-l1-2-3-treble-a-b-dev...   
28       /t/app-8-0-skip-track-by-volume-keys.3914337/   
29   /t/app-microg-gmscore-lightweight-free-softwar...   
30   /t/lawnchair-customizable-pixel-launcher.3627137/   
31   /t/app-2-1-15-jun-2-1-better-wifi-on-off.1801284/   
32   /t/app-root-shizuku-10-11-feb-darq-per-app-sel...   
33   /t/app-mod-netflix-kodi-edition-kodi-fork-with...   
34   /t/app-root-snapchat-dark-mode-enable-dark-mod...   
35   /t/port-app-5-0-06-06-21-samsung-music-port-fo...   
36   /t/app-netflix-hd-for-widevine-l1-device.3942467/   
37           /t/black-hole-a-music-player-app.4430223/   
38   /t/app-8-0-open-source-root-oandbackupx-apps-a...   
39   /t/app-4-0-3-gplv3-apkupdater-2-get-updates-fo...   
40   /t/tool-merjapp-merge-whatsapp-databases.2896496/   
41   /t/app-5-1-powertube-fast-video-downloader-bas...   
42   /t/free-open-source-keyboard-button-mapper-no-...   
43   /t/app-6-0-rethinkdns-anti-censorship-adblocke...   
44   /t/app-2-3-droidshows-a-reboot-of-droidseries-...   
45   /t/app-2-2-drivedroid-host-iso-img-files-to-bo...   
46   /t/app-mod-cast-jiotv-mod-with-miracast-androi...   
47          /t/app-7-foss-browser-open-source.3500091/   
48   /t/app-4-4-download-navi-open-source-download-...   
49    /t/launcher-2-2-t-ui-linux-cli-launcher.3146288/   
50   /t/app-ddc-toolbox-create-vdcs-for-viper-and-j...   
51   /t/app-4-4-open-source-andotp-open-source-two-...   
52   /t/app-6-0-code-scanner-a-foss-android-app-for...   
53               /t/app-7-0-smart-autoclicker.4379467/   
54   /t/app-mod-nonroot-songtube-modded-youtube.428...   
55   /t/tool-mobydroid-android-device-manager-for-l...   
56   /t/tool-whatsapp-key-db-extractor-crypt6-12-no...   
57   /t/app-4-1-contact-diary-keep-track-of-the-peo...   
58             /t/app-desktop-android-toolbox.4401109/   
59   /t/app-5-1-root-disabler-app-to-easily-disable...   
60   /t/app-5-0-smbsync2-one-way-synchronization-ba...   
61                       /t/phoneprofilesplus.3799429/   
62   /t/app-5-0-root-prefedit-2-0-view-edit-shared-...   
63   /t/app-2-1-cpu-spy-v0-4-0-beta-view-time-in-st...   
64        /t/app-omnijaws-weather-fixed-v-1-1.4173199/   
65   /t/app-4-2-free-materialfbook-minimalist-faceb...   
66   /t/app-open-source-feedtv-news-and-tv-from-any...   
67   /t/app-2-3-gplv3-apkgrabber-re-release-of-apku...   
68   /t/app-android-5-0-rywall-uhd-wallpaper-app.

In [114]:
pages.to_csv("merged_pages.csv", sep=";")